contracts/router/Adapter.sol

```diff
@@ -0,0 +1,249 @@
+     */
+
+    function depositAddress(address _tokenIn, address _tokenOut)
+        external
+        view
+        returns (address)
+    {
+        return _depositAddress(_tokenIn, _tokenOut);
+    }
+
+    /**
+     * @notice Get query for a swap through this adapter
+     *
+     * @param _amountIn input amount in starting token
+     * @param _tokenIn ERC20 token being sold
+     * @param _tokenOut ERC20 token being bought
+     */
+    function query(
+        uint256 _amountIn,
+        address _tokenIn,
+        address _tokenOut
+    ) external view returns (uint256) {
+        if (_amountIn == 0 || !_checkTokens(_tokenIn, _tokenOut)) {
+            return 0;
+        }
+        return _query(_amountIn, _tokenIn, _tokenOut);
+    }
+
+    /**
+     * @notice Execute a swap with given input amount of tokens from tokenIn to tokenOut,
+     *         assuming input tokens were transferred to depositAddress(_tokenIn, _tokenOut)
+     *
+     * @param _amountIn input amount in starting token
+     * @param _tokenIn ERC20 token being sold
+     * @param _tokenOut ERC20 token being bought
+     * @param _to address where swapped funds should be sent to
+     *
+     * @return _amountOut amount of _tokenOut tokens received in swap
+     */
+    function swap(
+        uint256 _amountIn,
+        address _tokenIn,
+        address _tokenOut,
+        address _to
+    ) external returns (uint256 _amountOut) {
+        require(_amountIn != 0, "Adapter: Insufficient input amount");
+        require(_to != address(0), "Adapter: Null address receiver");
+        require(_checkTokens(_tokenIn, _tokenOut), "Adapter: unknown tokens");
+        _approveIfNeeded(_tokenIn, _amountIn);
+        _amountOut = _swap(_amountIn, _tokenIn, _tokenOut, _to);
+        emit AdapterSwap(_tokenIn, _tokenOut, _amountIn, _amountOut);
+    }
+
+    // -- INTERNAL FUNCTIONS
+
+    /**
+     * @notice Return expected funds to user
+     *
+     * @dev this will do nothing, if funds need to stay in this contract
+     *
+     * @param _token address
+     * @param _amount tokens to return
+     * @param _to address where funds should be sent to
+     */
+    function _returnTo(
+        address _token,
+        uint256 _amount,
+        address _to
+    ) internal {
+        if (address(this) != _to) {
+            IERC20(_token).safeTransfer(_to, _amount);
+        }
+    }
+
+    /**
+     * @notice Check allowance, and update if it is not big enough
+     *
+     * @param _token token to check
+     * @param _amount minimum allowance that we need
+     * @param _spender address that will be given allowance
+     */
+    function _checkAllowance(
+        IERC20 _token,
+        uint256 _amount,
+        address _spender
+    ) internal {
+        uint256 _allowance = _token.allowance(address(this), _spender);
+        if (_allowance < _amount) {
+            // safeApprove should only be called when setting an initial allowance,
```

Comment:
`safeApprove` is also deprecated, why not just call `safeIncreaseAllowance`?

Reply:
The actual approval is supposed to be done once per token (set it to UINT_MAX and then it's never updated again). But this needs to work, if for whatever reason the allowance for the token was previously set to some finite amount.

As this is the base contract, we can't assume finite allowances won't be ever used, and SafeERC20.safeApprove reverts when both old and new allowance are non-zero. Thus, the extra zero approval.

IMO using `safeIncreaseAllowance` every time should be less efficient than setting allowance to infinity.